ios/Exponent/Kernel/Core/EXKernel.m

```diff
@@ -215,34 +217,7 @@ - (void)_dispatchJSEvent:(NSString *)eventName body:(NSDictionary *)eventBody to
 
 - (NSDictionary *)initialAppPropsFromLaunchOptions:(NSDictionary *)launchOptions
```

Comment:
We'll need to update the docs so that people don't expect the root component's `props.exp` to contain the launch notification. (This has come up in bug reports / issues a few times.)

Reply:
I expect initial notification to appear in `props.exp`. 🙂 With `UNNotificationFramework` we can't use `launchOptions` to access the initial notification synchronously, but as soon as we set a delegate on `UNNotificationCenter.defaultCenter`, the delegate is called with `userNotificationCenter:didReceiveNotificationResponse:`. Our delegate is `EXUserNotificationManager`, which, if the kernel couldn't send the notification to a running application, saves the notification in `<experienceId, pendingNotification>` dictionary. When the app starts to run, `EXReactAppManager` asks `EXUserNotificationManager` for initial notification for given `experienceId`.

Notification could be lost if `EXReactAppManager` would ask `EXUserNotificationManager` for pending notification before `UNNotificationCenter` would notify delegate of the notification response. I can't say I'm 100% sure this will never happen, but as the Apple documentation states:

> Notifications that cause your app to be launched are delivered shortly after these methods ([`application:didFinishLaunchingWithOptions:`]) finish executing.

Reply:
OK, thanks for the explanation.

> Notification could be lost if EXReactAppManager would ask EXUserNotificationManager for pending notification before UNNotificationCenter would notify delegate of the notification response.

In this scenario -- if it happens -- would we send the notification to the app when it subscribes to new notification events? We may still need to save up pending notifications after an app has launched to make sure we handle this scenario:

1. App launches. `EXReactAppManager` finds no pending notification from `EXUserNotificationManager`.
2. iOS invokes `userNotificationCenter:didReceiveNotificationResponse:`. This could happen either because (a) the app was launched from a notification and iOS delivered it late or (b) the app received a push notification after the app launched.
3. App JS subscribes to notification events.
4. How do we get the notification from step 2?

Reply:
Reading the docs more, it makes me think that:

1. For now, we should keep the behavior in this PR. Apple's docs make it sound like they invoke the delegate method very soon after the app launches since they want the notification delegate to be set up before `application:didFinishLaunchingWithOptions:` returns. This will likely behave correctly for Expo's current API of providing the launch notification in `props.exp`.

2. In the future (this PR certainly has enough changes!) I would like to feel even more confident in our notification API and defend against changes in Apple's behavior by changing Expo's API to be closer to Apple's. There are two potential race condition bugs I see. One is that the launch notification gets sent to the notification delegate after the JS app launches (too late to pass it in through `props.exp`). The second is that after the native app launches, we may receive a notification response before the JS app launches if the timing is just right. In this case, we'd possibly (a) have two pending notifications if the app was launched from a different notification or (b) incorrectly think the notification response was used to launch the app and store it as a pending notification that gets passed to the JS app through `props.exp`.

To address (2), I think we'd need to provide an API similar to Apple's. Specifically, instead of passing notifications through `props.exp` we'd always pass notifications through the event emitter. (Apple removed `launchOptions` and now always uses the delegate method => Expo removes `props.exp.notification` and now always uses the event listener.)

Also similar to Apple's API, we'd require Expo developers to subscribe to the notification listener in JS before a cutoff point that we choose (for example, before `Expo.registerComponent` is called?)  -- if they do this, we guarantee that the notification used to launch the app (if the app was launched from a notification) will get sent to the listener. We can implement this by buffering up **all** pending notifications the native app receives before the cutoff point (ex: `Expo.registerComponent()`) and then replaying all of them by emitting events. This way the app is guaranteed to receive the notification that launched the app and will also receive all notifications that were delivered after app launched -- perfect delivery.